[scheduler/cuebot/pycue/rqd/pyoutline] Booking by Slot

DiegoTavares opened the pull request:
Add a new booking mode that doesn't take cores and memory into consideration, but a predefined limit on how many concurrent frames a host is allowed to run.

Rationale: Booking by slot is useful for pipelines where frames are small and limited not by their
cpu/memory consumption but by other resources like storage bandwidth or network availability. In
these scenarios, limiting the concurrency is more important than the resource consumption.

**Attention:*** This branch is stacked on top of https://github.com/AcademySoftwareFoundation/OpenCue/pull/2002

Tasks:
 - [x] Implement booking logic on Scheduler
 - [x] Add new columns to Host to mark how many slots are available and fill them up on Cuebot
 - [x] Add new column to Layer to define slot limit and implement logic to fill it up on Cuebot
 - [x] Handle new attributes on Host and Layer using pycue
 - [x] Handle new attributes on Host and Layer using pyoutline
 - [x] Allow setting hosts' slot limit on cuegui
 - [x] Change rqd to ignore core count for slot base booking
 - [x] Change scheduler to mark slot layers as non-threadable

## Cuegui Changes

Add a new column to HostMonitorTree to show the number of concurrent slots:
<img width="1750" height="236" alt="Screenshot 2025-12-18 at 10 34 19 AM" src="https://github.com/user-attachments/assets/8396e0a3-afd7-4d81-af9d-5d285937e4fd" />

Add a new menu option on HostMonitorTree to change the concurrent slots limit value:
<img width="207" height="288" alt="Screenshot 2025-12-18 at 10 34 26 AM" src="https://github.com/user-attachments/assets/2dbf199e-93d4-47df-af12-0d69e2863d74" />
<img width="569" height="223" alt="Screenshot 2025-12-18 at 10 37 07 AM" src="https://github.com/user-attachments/assets/37533ad8-7808-488f-8df4-88b0ba57086b" />

## Attention

The new feature is initially being developed on top of the new scheduler module. At this time Cuebot will not take slot booking into consideration. A future PR will handle implementing the Cuebot changes to avoid touching too many files on a single PR.

DiegoTavares committed:
Add int_concurrent_frames_limit to host

This field limits the number of concurrent frames allowed to run on a specific host.

DiegoTavares committed:
Add Layer Slots Required Field

DiegoTavares committed:
[rework] Handle hardcoded values

DiegoTavares committed:
[scheduler/cuebot] Introduce booking by slot to scheduler

This commit is the first step towards the goal of allowing a new booking mode that doesn't take
cores and memory into consideration, but a predefined limit on how many concurrent frames a host is
allowed to run.

Rationale: Booking by slot is useful for pipelines where frames are small and limited not by their
cpu/memory consumption but by other resources like storage bandwith or network availability. In
these scenarios, limiting the concurrency is more important than the resource consumption.

DiegoTavares committed:
Add slots_required field to layer in OpenCue DTD and PyOutline layer

Add slots_required attribute to layer for slot-based booking

DiegoTavares committed:
Clean up warning from rust modules

DiegoTavares committed:
[pycue/cuegui/cuebot] Add concurrent_procs_limit

Add a menu action for setting a host's slot limit.
When a limit is defined, booking will only allocate layers with slots_required > 0 to be executed on this host. Which means regular booking by cores/memory/gpu becomes disabled.
(0 for no limit, >0 for specific limit)

Changes:
 - Add new proto field to Host and NestedHost
 - Change pycue to allow setting concurrent_procs_limit
 - Change cuegui action menu to add an option to update the new field
 - Update Cuebot to receive the request and update the database

DiegoTavares committed:
Rename concurrent_procs to concurrent slots

DiegoTavares committed:
Minor fixes

DiegoTavares committed:
Merge branch 'master' into slot-based-scheduling

DiegoTavares committed:
Revert re-formatting

DiegoTavares committed:
Fix formatting

DiegoTavares committed:
Fix formatting

DiegoTavares committed:
Add slots_required to rqd.RunFrame

DiegoTavares committed:
Implement slot based booking on rqd

DiegoTavares committed:
spotless Apply

DiegoTavares committed:
Add slots_required to LayerMonitorTree

DiegoTavares committed:
Add slots required configuration to LayerDialog

DiegoTavares committed:
Merge branch 'master' into slot-based-scheduling

DiegoTavares committed:
Update documentation

DiegoTavares committed:
Version Up

DiegoTavares committed:
Fix migration id

DiegoTavares committed:
Fix python linting

ramonfigueiredo commented:
@DiegoTavares 
FYI ... I am review this PR

ramonfigueiredo commented on rust/crates/rqd/src/system/machine.rs:
```diff
@@ -711,6 +816,38 @@ impl Machine for MachineMonitor {
+
+        match slot_state.as_mut() {
+            Some(slot_state) => {
+                if slot_state.slots_consumed + requested_slots < slot_state.slot_limit {
```
```suggestion
                if slot_state.slots_consumed + requested_slots <= slot_state.slot_limit {
```

ramonfigueiredo commented on proto/src/report.proto:
```diff
@@ -183,3 +186,12 @@ message RqdReportStatusRequest {
+}
+
+message RqdReportGetHostSlotsLimitResponse {
+    int64 slots_limit = 1;
```
```suggestion
    int32 slots_limit = 1;
```

ramonfigueiredo replied:
Issue: `RqdReportGetHostSlotsLimitResponse` uses `int64` `slots_limit` while all other slot fields use `int32`

Impact: Type conversion issues between systems

Required fix: 
- Change to `int32` for consistency

ramonfigueiredo commented on cuebot/src/main/java/com/imageworks/spcue/dao/postgres/LayerDaoJdbc.java:
```diff
@@ -571,6 +573,12 @@ public void updateTimeoutLLU(LayerInterface layer, int timeout_llu) {
     }
 
+    @Override
+    public void updateLayerSlotsRequired(LayerInterface layer, int slots) {
```
```suggestion
    public void updateLayerSlotsRequired(LayerInterface layer, int slots) {
        if (slots < 0) {
            throw new IllegalArgumentException("Slots required cannot be negative");
        }
```

ramonfigueiredo commented on proto/src/host.proto:
```diff
@@ -273,7 +276,12 @@ message Host {
+    
+    // When a limit is defined, booking will only allocate layers with slots_required > 0 to be
+    // executed on this host. Which means regular booking by cores/memory/gpu becomes disabled.
+    // (0 for no limit, >0 for specific limit)
```
```suggestion
    // Controls slot-based concurrent frame limiting on this host.
    // 0: slot-based booking is disabled (default behavior using cores/memory/gpu)
    // >0: slot-based booking is enabled with specified limit (only layers with slots_required > 0)
```

ramonfigueiredo replied:
Default value inconsistency:

Issue: Proto documentation, database default, and feature docs are inconsistent about what 0 vs -1 means.

Impact: Ambiguous semantics - code must handle both -1 and 0 as "disabled"

Required Fix: Standardize on one value (recommend 0) and update docs consistently

ramonfigueiredo reviewed (commented):
Hi @DiegoTavares

Some recommended changes are below.

Thanks for your contribution!

ramonfigueiredo commented:
### Missing Slot reporting from RQD to Cuebot

Location: rust/crates/rqd/src/system/machine.rs:932-969

Issue: RQD tracks slots internally but does NOT report slot consumption to Cuebot. The HostReport proto contains CoreDetail and RenderHost but neither has slot fields.

Impact:
- Cuebot cannot make informed booking decisions without knowing available slots
- Scheduler's host cache has no slot visibility
- Multiple dispatchers could over-book hosts

Root Cause: proto/src/report.proto missing slot fields in CoreDetail (lines 39-49) or RenderHost (lines 70-91)

Required Fix:

1. Extend CoreDetail message with: int32 total_slots, int32 idle_slots, int32 consumed_slots
2. Update RQD's collect_host_report() to populate these fields
3. Update Cuebot's report handler to process slot metrics

ramonfigueiredo commented on cuebot/src/main/java/com/imageworks/spcue/dao/postgres/HostDaoJdbc.java:
```diff
@@ -562,6 +565,23 @@ public void updateThreadMode(HostInterface host, ThreadMode mode) {
     }
 
+    @Override
+    public void updateConcurrentSlotsLimit(HostInterface host, int limit) {
```
Issue: No validation for negative values or maximum reasonable limits

DiegoTavares replied:
Not really a problem. Any negative value will have the same behavior of -1 since <0 means no limit.

ramonfigueiredo commented:
### No cross-validation between host and layer settings

Locations: Throughout cuebot/cuegui

Issue: No check that `layer.slots_required <= host.concurrent_slots_limit`

Impact: Could result in layers that can never run

Suggested fix: Add validation in dispatcher or show warnings in GUI

ramonfigueiredo commented:
### Missing Slot utilization monitoring

Location: cuegui/cuegui/HostMonitorTree.py

Issue: No column showing "Slots In Use" vs "Slots Available" for real-time monitoring

Suggested Fix: Add columns showing `running_procs_count` and available slots

ramonfigueiredo commented on cuebot/src/main/resources/conf/ddl/postgres/migrations/V38__Add_host_frame_limit.sql:
```diff
@@ -0,0 +1,7 @@
+-- Add a field to limit the max amount of concurrent frales a host can run
+-- -1 means no limit
+alter table host
+    add int_concurrent_slots_limit INT NOT NULL DEFAULT -1;
```
Issue: Filename says "frame_limit" but adds `concurrent_slots_limit` field

Suggested Fix: Rename to `V38__Add_host_concurrent_slots_limit.sql`

ramonfigueiredo commented:
## Other minor changes

### Inconsistent Terminology in GUI Tooltips
Locations:
- cuegui/cuegui/HostMonitorTree.py:167 - "Usually: 1 frame = 1 slot"
- cuegui/cuegui/MenuActions.py:1950 - "usually a frame consumes 1 slot"

Suggested fix: Use consistent phrasing

### Column Display Inconsistency

Locations:
- HostMonitorTree.py:166 - Shows "-" for values < 0
- LayerMonitorTree.py:167 - Shows "-" for values <= 0

Suggested fix: Use consistent threshold logic

### Missing Docstring Type Annotation

Location: pycue/opencue/wrappers/host.py:131-132

Issue: setConcurrentSlotsLimit() lacks :type: and :param: docstring

Suggested fix: Add type annotations like setSlotsRequired() has

### Dummy Cuebot Always Returns -1

Location: rust/crates/dummy-cuebot/src/report_servant.rs:84

Issue: Hardcoded -1 prevents testing slot-based booking with dummy-cuebot

Suggested fix: Make configurable for testing

### Deprecated Field Usage

Location: rust/crates/scheduler/src/pipeline/dispatcher/actor.rs:973

Issue: slots_required: 0 hardcoded in deprecated field

Note: May be intentional for backward compatibility

DiegoTavares committed:
Fix review comments

DiegoTavares committed:
Improve required_slots accounting

Cuebot now accounts for a sum of frames' consumed slots, not 1 slot per frame as previously.

DiegoTavares committed:
Add runningSlots tracking to Host and display available slots in GUI

- Add runningSlots field to Host protobuf and HostEntity
- Update Whiteboard DAO queries to include runningSlots
- Show available slots (concurrentSlotsLimit - runningSlots) in CueGUI host monitor tree

DiegoTavares committed:
Fix cicd

DiegoTavares committed:
Fix tests and spotless errors

ramonfigueiredo reviewed (approved):
LGFM. Approved!

Thanks @DiegoTavares

ramonfigueiredo commented on cuebot/src/main/java/com/imageworks/spcue/servant/ManageLayer.java:
```diff
@@ -432,6 +434,15 @@ public void setTimeoutLLU(LayerSetTimeoutLLURequest request,
+        jobManager.setLayerSlotsRequired(layer, request.getSlots());
+        responseObserver.onNext(LayerSetSlotsRequiredResponse.newBuilder().build());
+        responseObserver.onCompleted();
+    }
```
`ManageLayer.setSlotsRequired` skips `attemptChange` guard

cuebot/.../servant/ManageLayer.java:437-444

Every other layer setter (`setThreadable`, `setTimeout`, `setMinCores`, etc.) uses `attemptChange(env, property, ...)` to prevent modifications to finished jobs. `setSlotsRequired` does not, allowing modification of finished jobs when `layer.finished_jobs_readonly` is configured.

ramonfigueiredo commented on cuebot/src/main/java/com/imageworks/spcue/dao/postgres/HostDaoJdbc.java:
```diff
@@ -395,22 +401,23 @@ public CallableStatement createCallableStatement(Connection con) throws SQLExcep
             + "  ts_ping = current_timestamp, "
-            + "  str_os = ? "
+            + "  str_os = ?, "
+            + "  int_running_slots = ? "
```
Race condition on slot overbooking

The `runningSlots` value is set from host reports (periodic). Concurrent dispatcher threads can read the same stale count and both book a frame, exceeding the limit. This needs database-level locking (SELECT ... FOR UPDATE) or atomic compare-and-update.

ramonfigueiredo commented on rust/crates/rqd/src/frame/running_frame.rs:
`clone_into_running_frame_info` doesn't populate `slots_required`

rust/crates/rqd/src/frame/running_frame.rs:1247-1281 : Uses `..Default::default()` which sets `slots_required: 0`. The scheduler won't know slot requirements from running frame reports.

ramonfigueiredo commented on cuebot/src/main/java/com/imageworks/spcue/service/HostManagerService.java:
```diff
@@ -93,6 +93,17 @@ public void setHostFreeTempDir(HostInterface host, Long freeTempDir) {
     }
 
+    @Override
+    public void setConcurrentSlotsLimit(HostInterface host, int limit) {
```
Missing input validation

- `HostManagerService.setConcurrentSlotsLimit` : passes directly to DAO without bounds checking

ramonfigueiredo commented on cuebot/src/main/java/com/imageworks/spcue/dispatcher/HostReportHandler.java:
```diff
@@ -185,13 +185,21 @@ public void handleHostReport(HostReport report, boolean isBoot) {
+             * Verify all the frames in the report are valid. Frames that are not valid are removed.
+             */
+            List<RunningFrameInfo> runningFrames = verifyRunningFrameInfo(report);
+            int host_running_slots =
```
HostReportHandler.java:193

`host_running_slots` uses `snake_case`; should be hostRunningSlots.

 It's standard Java convention to use `camelCase` for local variables, but it's a stylistic preference, not a  
  bug.

ramonfigueiredo commented on docs/_docs/concepts/glossary.md:
```diff
@@ -123,8 +123,19 @@ type of job to the cue, including Maya, Katana, or even shell commands.
+A booking mode where hosts are scheduled by slots instead of cores, memory, or
+GPU resources. It is enabled by setting a concurrent slots limit on hosts and a
+`slots_required` value on layers.
 
```
Minor detail:

Glossary entries break alphabetical order : "Slot" placed between "Proc" and "OpenCue REST Gateway"

ramonfigueiredo commented on cuebot/src/main/resources/conf/spring/applicationContext-grpc.xml:
```diff
@@ -14,12 +14,12 @@
-       xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-       xsi:schemaLocation="
-            http://www.springframework.org/schema/beans http://www.springframework.org/schema/beans/spring-beans-2.5.xsd">
+<beans
```
Spring XML reformatting adds diff noise

applicationContext-grpc.xml : Extensive reformatting unrelated to the feature.

DiegoTavares committed:
Merge branch 'master' into slot-based-scheduling

DiegoTavares committed:
Fix issue identified on the last review pass

DiegoTavares commented:
Branch has been reworked into #2486
